### DISCUSSION TITLE
How to make the graphics scale and look nice?

### DISCUSSION THREAD

**Discussion by apcsb** [ihhub/fheroes2#8635] [Q&A] [answered]
Hi. I have just downloaded fHeroes2 and looked at a number of issues and discussions. The challenge I have is as follows:
I would prefer to play the game as follows:
1. Game fills the entirety of the screen (standard 1920x1080 FullHD 16:9 screen)
2. Fullscreen
3. Does not f*ck up my Windows resolution (example here #7210 )
4. City/Battle screens fill the entire screen
5. Fonts are readable (i.e. not too small)

**Question 1: this is ever possible?**

So far I have the following:
* Playing in any resolution aside from 640x480 or 640x512 makes Castle/Battle interfaces pop up in a window, which ruins the immersion (is there a problem to scale the bitmaps by default)?
* Playing in any other resolution than native in fullscreen messes up my other windows, since the game switches the "native resolution" of the entire Windows stack, instead of creating a viewport using the native resolution and scaling everything inside it (see #7210)
* I can't seem to figure out how fheroes2.cfg file with settings referenced in #7210 and it doesn't seem to do anything. I can't see any references to this file in the source code. Is it still supported? Is it documented somewhere?
* If playing with the native resolution, the fonts are too small to read for me (15" screen). Can I scale just the fonts somehow? I've found some settings like "**fonts normal size**" for the CFG file, but they don't seem to do anything.
*  If playing in Window mode with, say 640x480 resolution and maximizing the window, I get black borders, which look ugly.

So basically, one can either have immersion, or functioning Windows or black borders, but not everything at once.

**Question 2: what can be done to bring the game as close to the optimal settings as possible?**
If someone/multiple people bother to explain this to me, I commit to contribute to a Wiki page with a proper doc on this matter, based on their contributions.

**Question 3: am I understanding correctly, that the engine just does not have a "flexible" bitmap scaler to scale the OG 640x480 assets for *any* resolution?**
I've found a bunch of discussion and issues that suggest that something like this exists, but it doesn't seem to work. Or is there another hidden option somewhere? 

Thanks!

**Comment by oleg-derevenetz** [comment, marked as answer]
> Playing in any resolution aside from 640x480 or 640x512 makes Castle/Battle interfaces pop up in a window, which ruins the immersion (is there a problem to scale the bitmaps by default)?

Original battle window resources are designed for 640x480 resolution (i.e. 4:3), so if you have 16:9 resolution (like your 1920x1080) there is just no way to scale them to cover the entire window without introducing distortions (stretching in the horizontal direction will be greater than stretching in the vertical direction). Hence the need for a window.

> If playing with the native resolution, the fonts are too small to read for me (15" screen). Can I scale just the fonts somehow?

You should be able to play in the `960x540 (x2.0)` resolution with scaling, so fonts will be big enough. If you want to use non-hardware resolution change method on Windows, you should also set `cursor soft rendering = on` in the `fheroes2.cfg` (see below).

> I've found some settings like "fonts normal size" for the CFG file, but they don't seem to do anything.

There have been no such settings for a long time. They were once used to support third-party TTF fonts, but this support has been gone for a long time.

> I can't seem to figure out how fheroes2.cfg file with settings

It is located in your `%APPDATA%\fheroes2` directory.

**Reply by apcsb** [reply]
Hi, thanks for answering! 
I've figured as much and after some googling found a reference to %APPDATA% at Reddit. I'll try to pen up an article and see if I can submit a PR to get it into a Wiki.

As for the "Original battle window resources are designed for 4:3". I get it, yet I think there are some low-hanging fruits here. In the order of increasing complexity (based on my perception, may be wrong - you'd know better :) ):

1. Somehow, for instance, the DOSBOX GOG version makes it look OK even on 16:9 stretching it for full screen. Having a "Just stretch it, I don't care" option in the CFG file would be the easiest way to win. "UI screens scaling = fill"

2. If someone does not like the stretched look  (or has different screen proportions) one could simply stretch the battle/town screens (while preserving proportions) to fill the max vertical space and create some UI frames for the left/right margins so that the underlying main screen UI is not visible (I keep trying to click the town list from the town screen). Basically, the same letterboxing as today, but instead of ugly black margins - nice looking UI frames. "UI screens scaling  = fit". Here's a quick mockup ( I would fill the blank areas with some ornament, like it was common in the medieval books, but I couldn't quickly find anything that would match visually, so it's not there).
![image](https://github.com/ihhub/fheroes2/assets/5718377/3a0d56e9-ccea-4e70-bb5e-b92189f40738)

3. For the battle screen one could simply add more grass/dirt/etc on the sides and reposition the heroes slightly more towards the edges accordingly, leaving the battle grid as is in the center (scaled to fill all the vertical space). Don't know how complex it is to implement, but I assume that the battlefield is generated dynamically anyway, or isn't it?
Here's a (pretty ugly) mockup, but I hope it conveys the idea
![image](https://github.com/ihhub/fheroes2/assets/5718377/f269a53b-de05-4bed-aa5a-4e7e4083769f)


4. Build new UI layouts. Here are some examples.
4.1 Battle screen with original 4:3 layout incorporated, logs/buttons/ info popups moved (forgot the scrollbar for the logs..)
![image](https://github.com/ihhub/fheroes2/assets/5718377/65af3302-20c6-44e2-aad2-3a9f22682151)

4.2 Town UI incorporating the 4:3 screen
![image](https://github.com/ihhub/fheroes2/assets/5718377/3159af7b-41a2-4990-8819-c12426ad9de0)

4.3 Town UI incorporating just the background (the background proportions are the same as in the original 4:3 screen)
![image](https://github.com/ihhub/fheroes2/assets/5718377/500ddf89-76a6-4261-8e9c-c12bbbbc6226)

![image](https://github.com/ihhub/fheroes2/assets/5718377/ed640e24-2824-421a-9ae3-c27e7f18a480)

The other screens (spellbook, Hero Screen, Hero exchange screen  (army/artifacts), to be honest, don't bother as much, but could also potentially be implemented using methods 1 and 2.

What do you think?

_P.S. The first time I saw the gems in the Adventure Screen frame, I assumed they would indicate the days of the week and weeks of the month, so that one doesn't have to click that small bottom left window all the time.. :)

**Reply by oleg-derevenetz** [reply]
> What do you think?

At the time, it was decided to follow the path of HoMM3 HotA. HotA developers faced essentially the same issue: original HoMM3 resources were designed for 4:3 resolution (800x600), so they used the same "window" approach:

![hota](https://github.com/ihhub/fheroes2/assets/32623900/43afae55-c43b-4b1b-a9b1-318780d5c651)

To redo the layouts so that it looks elegant, and not crude and shabby, we need to have an artistic taste. The developers of HotA did not dare to do this, despite the fact that, unlike us, they definitely have artists (all these new factions with their new units and castles).

> so that one doesn't have to click that small bottom left window all the time..

In fheroes2, you don't need to click that window in higher resolutions (e.g. 1024x768), there is enough space in these resolutions to show everything at once (date, army and resources).

**Reply by apcsb** [reply]
> In fheroes2, you don't need to click that window in higher resolutions
Yeah, but it higher resolutions the battle window is too small. From what I read in discussions and forums, it is recommended to stick as close to the original 640x480 (or use x2/x2.2 upscaling) as possible.

As for the artistic taste - I think you got it quite well with the picture frames, and this is pretty much all one needs. The rest can be produced by AI  - there are many tools that allow you to upload an image and say "Create XYZ styled like the uploaded image".
